uniswap_edpr/src/logics/shared_algorithm_lens/src/lib.rs

```diff
@@ -0,0 +1,86 @@
+use shared_algorithm_lens_accessors :: * ;
+
+# [derive (Clone , Debug , Default , candid :: CandidType , serde :: Deserialize , serde :: Serialize)]
+pub struct LensValue {
+  pub address: String,
+  pub current_price: f32,
+  pub range_top: f32,
+  pub range_bottom: f32,
+  pub edpr: f32
+}
+
+pub async fn calculate (targets : Vec < String >) -> LensValue {
+  let pool_fees_result = get_get_last_snapshot_value_in_pool_fees (targets . get (0usize) . unwrap () . clone ()) . await ;
+  let tc28x6_result = get_get_last_snapshot_value_in_tcumul_28x6hr (targets . get (1usize) . unwrap () . clone ()) . await ;
+  let v3pool_result = get_get_last_snapshot_value_in_v3pool (targets . get (2usize) . unwrap () . clone ()) . await ;
+  let eth_usdc_price_result = get_get_last_snapshot_value_in_eth_usdc_price (targets . get (3usize) . unwrap () . clone ()) . await ;
+  
+  let fees_24h_usd = pool_fees_result.clone().unwrap().result[0].fees_24h_usd;
+  let t0_decimals = pool_fees_result.clone().unwrap().result[0].pool_summary_level_1.t0_decimals;
+  let t1_decimals =  pool_fees_result.clone().unwrap().result[0].pool_summary_level_1.t1_decimals;
+
+  let eth_price_usd  = eth_usdc_price_result.clone().unwrap().result.t1_price_usd;
+
+  let address = v3pool_result.clone().unwrap().result.address;
+  let current_tick_liquidity_str = v3pool_result.clone().unwrap().result.liquidity;
+  let current_tick_liquidity = i128::from_str_radix(current_tick_liquidity_str.trim_start_matches("0x"), 16).unwrap();
+  let sqrt_ratio_x96_str = v3pool_result.clone().unwrap().result.sqrt_ratio_x96;
+  let sqrt_ratio_x96 = u128::from_str_radix(sqrt_ratio_x96_str.trim_start_matches("0x"), 16).unwrap();
+  let tick_current = v3pool_result.clone().unwrap().result.tick_current;
+  let tick_spacing = v3pool_result.clone().unwrap().result.tick_spacing;
+  let token0 = v3pool_result.clone().unwrap().result.token0;
+
+  let tick_cumul_28x6h = tc28x6_result.unwrap().0;
+
+  let current_sqrt_price = sqrt_ratio_x96 as f32 / f32::powf(2.0,96.0);
+  let mut current_price = current_sqrt_price * current_sqrt_price;
+
+  let mut compressed: i32 = tick_current / tick_spacing;
+  if tick_current <0 && tick_current % tick_spacing != 0 {
+    compressed -= 1;
+  }
+  let floor = compressed * tick_spacing;
+  let mut min_tick = floor;
+  let mut max_tick = floor + tick_spacing;
```

Comment:
This part is calculated as follows

>While liquidity can only be added at initialized ticks, the current market-price can be between initialized ticks. In the example above, the USDC-WETH 5-bps pool is at tick 202919. This number is not evenly divisible by the tick-spacing of 10, we need to find the nearest numbers above and below that are. This corresponds to tick 202910 and tick 202920, which we showed in Figure 1.
https://blog.uniswap.org/uniswap-v3-math-primer#what-does-a-tick-represent

I thought you wanted to calculate the min tick and max tick of the tick range to which you belong.
I don't understand the intent of `compressed -= 1` in considering the case of indivisibility.
My calculation image is below.
```rust
let remainder = tick_current % tick_spacing;
let (min_tick, max_tick) = if remainder > 0 {
  let min = tick_current - remainder;
  let max = min + tick_spacing;
  (min, max)
} else {
  let min = tick_current - (tick_spacing - remainder);
  let max = min + tick_spacing;
  (min, max)
}
```

Reply:
Let's say this is your price (tick) range of interest (although we don't know the upper and lower boundaries yet):

![image](https://github.com/horizonx-tech/chainsight-showcase/assets/37236322/efde4668-75a1-4296-8064-b429e346991f)

Given the tick value representing the current price, we want to set up the initial min_tick and max_tick as follows:

![image](https://github.com/horizonx-tech/chainsight-showcase/assets/37236322/130e3851-1820-468c-a337-a01843f0e3aa)

Note that the tick value can be negative, in which case `compressed -= 1` is necessary in order to move min_tick to the lower edge of the tick, instead of the higher edge, which would otherwise be the result of the calculation because the higher edge is closer to 0.

Reply:
>Given the tick value representing the current price, we want to set up the initial min_tick and max_tick as follows:

I think I understand the situation here.

> While liquidity can only be added at initialized ticks, the current market-price can be between initialized ticks. In the example above, the USDC-WETH 5-bps pool is at tick 202919. This number is not evenly divisible by the tick-spacing of 10, we need to find the nearest numbers above and below that are. This corresponds to tick 202910 and tick 202920, which we showed in Figure 1.
https://blog.uniswap.org/uniswap-v3-math-primer#what-does-a-tick-represent

Take the above example,
tick_current = 202919
tick_spacing = 10

[1] Then I think max,min of intial are as follows.
min_tick = 202910
max_tick = 202920

If the above is true,
[2] It would be calculated as follows. I added the calculated values in the comments.

<details>
<summary>For positive values</summary>
```rust
let mut compressed: i32 = tick_current / tick_spacing; // 202919 / 10 = 20291.9
if tick_current <0 && tick_current % tick_spacing != 0 {
  compressed -= 1;
}
let floor = compressed * tick_spacing; // 202919
let mut min_tick = floor; // 202919
let mut max_tick = floor + tick_spacing; // 202929
```
</details>

<details>
<summary>For negative values</summary>
```rust
let mut compressed: i32 = tick_current / tick_spacing; // -202919 / 10 = -20291.9
if tick_current <0 && tick_current % tick_spacing != 0 {
  compressed -= 1; // -20291.9 - 1.0 = -20292.9
}
let floor = compressed * tick_spacing; // -202929
let mut min_tick = floor; // -202929
let mut max_tick = floor + tick_spacing; // -202919
```
</details>

Reply:
Remember that `:i32` makes `compressed` into an integer.

Positive tick:
```
  let tick_current = 202919;
  let tick_spacing = 10;
 
  let mut compressed: i32 = tick_current / tick_spacing; // 20291 (:i32 makes it an integer)
  if tick_current <0 && tick_current % tick_spacing != 0 {
    compressed -= 1;
  }
  let floor = compressed * tick_spacing; // 202910
  let mut min_tick = floor; // 202910
  let mut max_tick = floor + tick_spacing; // 202920
```

Negative tick:
```
  let tick_current = -202919;
  let tick_spacing = 10;
 
  let mut compressed: i32 = tick_current / tick_spacing; // -20291
  if tick_current <0 && tick_current % tick_spacing != 0 {
    compressed -= 1;                                     // -20292
  }
  let floor = compressed * tick_spacing; // -202920
  let mut min_tick = floor; // -202920
  let mut max_tick = floor + tick_spacing; // -202910
```

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=8afc618c3c028b5111dce12434e0cdce

Reply:
Sorry, I assumed it was a float. No problem.